### PR TITLE
[Core] Recursive Symlink Resolution

### PR DESCRIPTION
**Pull request by matekelemen**
Cherry-picked from #10004 at the request of @philbucher.

### CODE REVIEW

File: kratos/includes/kratos_filesystem.h [hunk 1]
```diff
@@ -98,6 +98,16 @@ std::vector<std::string> KRATOS_API(KRATOS_CORE) ListDirectory(const std::string
...
 
+/** @brief Resolve symlinks recursively.
+ *
+ *  @param rPath: path to a symbolic link.
+ *  @return The result of the recursive dereferencing.
+ *  @throws If the input path does not exist or the symlink is cyclic.
+ *  @note The existence of the final result is not checked and is up to the user.
+ *  @note The input is returned if it is not a symlink.
+ */
+std::string ResolveSymlinks(const std::string& rPath);
```
Comment (philbucher) [inline kratos/includes/kratos_filesystem.h]:
```suggestion
std::filesystem::path KRATOS_API(KRATOS_CORE) ResolveSymlinks(const std::filesystem::path& rPath);
```

will refactor the other methods also soon

File: kratos/sources/kratos_filesystem.cpp [hunk 2]
```diff
@@ -132,5 +133,23 @@ void MPISafeCreateDirectories(const std::string& rPath)
...
+
+std::string ResolveSymlinks(const std::string& rPath)
+{
+    std::filesystem::path path = rPath;
+    KRATOS_ERROR_IF_NOT(std::filesystem::exists(path)) << "File not found: " << rPath;
+
+    std::set<std::filesystem::path> symlinks;
+
+    while (std::filesystem::is_symlink(path))
+    {
```
Comment (philbucher) [inline kratos/sources/kratos_filesystem.cpp]:
```suggestion
    while (std::filesystem::is_symlink(path)) {
```

File: kratos/sources/kratos_filesystem.cpp [hunk 3]
```diff
@@ -132,5 +133,23 @@ void MPISafeCreateDirectories(const std::string& rPath)
     }
 }
 
+
+std::string ResolveSymlinks(const std::string& rPath)
+{
+    std::filesystem::path path = rPath;
+    KRATOS_ERROR_IF_NOT(std::filesystem::exists(path)) << "File not found: " << rPath;
+
+    std::set<std::filesystem::path> symlinks;
```
Comment (philbucher) [inline kratos/sources/kratos_filesystem.cpp:143]: why not unordered?

**Review by philbucher — COMMENTED** [review]
please also add 2 tests, one with and one without cyclic symlinks

File: kratos/sources/kratos_filesystem.cpp (hunk 3 above)
Comment (matekelemen) [reply kratos/sources/kratos_filesystem.cpp:143]: I don't expect many items in that set (probably in the single digits) so a hash table would be an overkill. In fact, I have a hard time imagining a scenario in which more than 2-3 indirections would be reasonable. A couple of lexicographical string comparisons in ```std::set``` surely outperform computing the hash of a lengthy absolute path.

**Comment by matekelemen** [comment]
@philbucher good thing you asked for the tests, ```std::filesystem``` is an inconsistent mess when it comes to following symlinks.

**Review by philbucher — DISMISSED** [review]
Perfect 👍🏻

**Comment by matekelemen** [comment]
fixed a small mistake in the tests; @philbucher can you please take a quick peek again?

**Review by philbucher — APPROVED** [review]

**Pull request merged**